ZelBack/src/services/fluxCommunicationMessagesSender.js

```diff
@@ -194,52 +299,73 @@ async function sendMessageToWS(message, ws) {
-    if (tempMesResponse) {
-      sendMessageToWS(tempMesResponse, ws);
+
+    const appsMessages = [];
+    if (!message || typeof message !== 'object' || typeof message.type !== 'string' || typeof message.version !== 'number'
+    || typeof message.broadcastedAt !== 'number' || !message.data) {
+      throw new Error('Invalid Flux App Request message');
+    }
+
+    if (message.version !== 1 && message.version !== 2) {
+      throw new Error(`Invalid Flux App Request message, version ${message.version} not supported`);
+    }
+
+    if (message.version === 1) {
+      if (typeof message.data.hash !== 'string') {
+        throw new Error('Invalid Flux App Request message, hash propery is mandatory on version 1');
+      }
+      appsMessages.push(message.data.hash);
+    }
+
+    if (message.version === 2) {
+      if (!message.data.hashes || !Array.isArray(message.data.hashes) || message.data.hashes.length > 500) {
+        throw new Error('Invalid Flux App Request v2 message');
+      }
+      for (let i = 0; i < message.data.hashes.length; i += 1) {
+        appsMessages.push(message.data.hashes[i]);
+        if (typeof hash !== 'string') {
+          throw new Error('Invalid Flux App Request v2 message');
+        }
+      }
+    }
+
+    const validTill = message.broadcastedAt + (65 * 60 * 1000); // 3900 seconds
+    if (validTill < Date.now()) {
+    // reject old message
       return;
     }
-    const appMessage = await appsService.checkAppMessageExistence(message.data.hash) || await appsService.checkAppTemporaryMessageExistence(message.data.hash);
-    if (appMessage) {
-      // const permanentAppMessage = {
-      //   type: messageType,
-      //   version: typeVersion,
-      //   appSpecifications: appSpecFormatted,
-      //   hash: messageHASH,
-      //   timestamp,
-      //   signature,
-      //   txid,
-      //   height,
-      //   valueSat,
-      // };
-      // a temporary appmessage looks like this:
-      // const newMessage = {
-      //   appSpecifications: message.appSpecifications || message.zelAppSpecifications,
-      //   type: message.type,
-      //   version: message.version,
-      //   hash: message.hash,
-      //   timestamp: message.timestamp,
-      //   signature: message.signature,
-      //   createdAt: new Date(message.timestamp),
-      //   expireAt: new Date(validTill),
-      // };
-      const temporaryAppMessage = { // specification of temp message
-        type: appMessage.type,
-        version: appMessage.version,
-        appSpecifications: appMessage.appSpecifications || appMessage.zelAppSpecifications,
-        hash: appMessage.hash,
-        timestamp: appMessage.timestamp,
-        signature: appMessage.signature,
-      };
-      myMessageCache.set(serviceHelper.ensureString(message), temporaryAppMessage);
-      sendMessageToWS(temporaryAppMessage, ws);
+
+    // eslint-disable-next-line no-restricted-syntax
+    for (const hash of appsMessages) {
```

Comment:
So this is a loop of up to 500 messages, Is it alright performance wise? Can we do just one search of database with all the messages avoiding this loop?

Reply:
I think this way it's fine, I have added a small delay between iterations.
It's also important that this way we are sending the messages one by one, and not flooding the node that asked for the messages with all the messages at the same time.